File: test_dhke/dhke.go
package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"encoding/hex"
	"math/big"
	"fmt"
	"io/ioutil"
	"log"
)

func loadRemotePublicKey() *ecdsa.PublicKey {
	remotePubKeyHex, err := ioutil.ReadFile("public_python")
	if err != nil {
		log.Fatal(err)
	}
	remotePubKeyBytes, err := hex.DecodeString(string(remotePubKeyHex))
	if err != nil {
		log.Fatal(err)
	}

	x := new(big.Int).SetBytes(remotePubKeyBytes[:32]) // X coordinate is the first 32 bytes
	y := new(big.Int).SetBytes(remotePubKeyBytes[32:]) // Y coordinate is the next 32 bytes

	return &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     x,
		Y:     y,
	}
}

func savePublicKey(privKey ecdsa.PrivateKey) {
	pubKeyBytes := append(privKey.PublicKey.X.Bytes(), privKey.PublicKey.Y.Bytes()...)
	pubKeyHex := hex.EncodeToString(pubKeyBytes)

	err := ioutil.WriteFile("public_go", []byte(pubKeyHex), 0644)
	if err != nil {
		log.Fatal(err)
	}
}

func generatePrivateKey() *ecdsa.PrivateKey {
	privKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		log.Fatal(err)
	}

	return privKey
}

func generateSharedSecret(remotePubKey ecdsa.PublicKey, privKey ecdsa.PrivateKey) *big.Int {
    sharedX, _ := elliptic.P256().ScalarMult(remotePubKey.X, remotePubKey.Y, privKey.D.Bytes())
    sharedSecret := new(big.Int).SetBytes(sharedX.Bytes())

	return sharedSecret
}

func saveSharedSecret(sharedSecret big.Int) {
    sharedSecretHex := hex.EncodeToString(sharedSecret.Bytes())
	err := ioutil.WriteFile("shared_secret_go", []byte(sharedSecretHex), 0644)
	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	remotePubKey := loadRemotePublicKey()
	fmt.Println("remotePubKey: ", remotePubKey)
	fmt.Printf("%T\n", remotePubKey)

	privKey := generatePrivateKey()
	fmt.Println("privKey: ", privKey)
    fmt.Printf("%T\n", privKey)

    savePublicKey(*privKey)

    sharedSecret := generateSharedSecret(*remotePubKey, *privKey)
    fmt.Println("sharedSecret:", sharedSecret)
	fmt.Printf("%T\n", sharedSecret)

	saveSharedSecret(*sharedSecret)  // for testing purposes only
}

File: test_dhke/dhke.py
import os
from ecdsa.curves import NIST256p
from ecdsa.ecdh import ECDH
from ecdsa.keys import VerifyingKey


def load_remote_public_key():
    remote_public_key_hex = read("public_go")
    return VerifyingKey.from_string(bytearray.fromhex(remote_public_key_hex), curve=NIST256p)


def generate_key_pair():
    ecdh = ECDH(curve=NIST256p)
    ecdh.generate_private_key()
    secret = ecdh.private_key
    public = ecdh.get_public_key()

    return secret, public


def write(path, data):
    with open(path, "w") as f:
        f.write(data)


def read(path):
    with open(path, "r") as f:
        return f.read()


def main():
    secret, public = generate_key_pair()
    write("public_python", public.to_string().hex())

    os.system("go run dhke.go")
    remote_pub_key = load_remote_public_key()

    ecdh = ECDH(curve=NIST256p, public_key=public, private_key=secret)
    ecdh.load_received_public_key(remote_pub_key)

    shared_secret_hex = ecdh.generate_sharedsecret_bytes().hex()
    write("shared_secret_python", shared_secret_hex)

    shared_secret_hex_go = read("shared_secret_go")
    assert shared_secret_hex == shared_secret_hex_go


if __name__ == "__main__":
    main()
